packages/app/src/cli/models/extensions/specifications/marketing_activity_extension_schemas/marketing_activity_extension_schema.ts

```diff
@@ -0,0 +1,187 @@
+
+const ScheduleFieldSchema = CommonFieldSchema.extend({
+  ui_type: zod.literal('schedule'),
+  use_end_date: zod.boolean(),
+})
+
+const ProductPickerFieldSchema = CommonFieldSchema.extend({
+  ui_type: zod.literal('product-picker'),
+  allow_product_image_selection: zod.boolean(),
+  allow_uploaded_image_as_product_image: zod.boolean(),
+  allow_free_image_as_product_image: zod.boolean(),
+  min_resources: zod.number().optional(),
+  max_resources: zod.number().optional(),
+  min_image_select_per_product: zod.number().optional(),
+  max_image_select_per_product: zod.number().optional(),
+})
+
+const SingleLineTextFieldSchema = CommonFieldSchema.extend({
+  ui_type: zod.enum(['text-single-line', 'text-email', 'text-tel', 'text-url']),
+  placeholder: zod.string().optional(),
+  min_length: zod.number(),
+  max_length: zod.number(),
+})
+
+const TextMultiLineFieldSchema = CommonFieldSchema.extend({
+  ui_type: zod.literal('text-multi-line'),
+  placeholder: zod.string(),
+  min_length: zod.number(),
+  max_length: zod.number(),
+})
+
+const DividerFieldSchema = BaseFieldSchema.extend({
+  ui_type: zod.literal('divider'),
+  title: zod.string(),
+  name: zod.string(),
+})
+
+const SelectFieldSchema = CommonFieldSchema.extend({
+  ui_type: zod.enum(['select-single', 'select-multiple']),
+  choices: zod.array(
+    zod.object({
+      label: zod.string(),
+      value: zod.string(),
+    }),
+  ),
+})
+
+const ParagraphFieldSchema = BaseFieldSchema.extend({
+  ui_type: zod.literal('paragraph'),
+  heading: zod.string().optional(),
+  body: zod.string().optional(),
+})
+
+const TypeAheadFieldSchema = CommonFieldSchema.extend({
+  ui_type: zod.literal('type-ahead'),
+  placeholder: zod.string(),
+})
+
+const NumberFieldSchema = CommonFieldSchema.extend({
+  ui_type: zod.enum(['number-float', 'number-integer']),
+  min: zod.number(),
+  max: zod.number(),
+  step: zod.number(),
+})
+
+const ImagePickerFieldSchema = CommonFieldSchema.extend({
+  ui_type: zod.literal('image-picker'),
+  min_resources: zod.number(),
+  max_resources: zod.number(),
+  allow_free_images: zod.boolean(),
+  alt_text_required: zod.boolean(),
+})
+
+const UISchemaMapping: {[key: string]: zod.Schema} = {
+  'budget-schedule': BudgetScheduleFieldSchema,
+  'discount-picker': DiscountPickerFieldSchema,
+  schedule: ScheduleFieldSchema,
+  'product-picker': ProductPickerFieldSchema,
+  'text-single-line': SingleLineTextFieldSchema,
+  'text-email': SingleLineTextFieldSchema,
+  'text-tel': SingleLineTextFieldSchema,
+  'text-url': SingleLineTextFieldSchema,
+  'text-multi-line': TextMultiLineFieldSchema,
+  'select-single': SelectFieldSchema,
+  'select-multiple': SelectFieldSchema,
+  paragraph: ParagraphFieldSchema,
+  'type-ahead': TypeAheadFieldSchema,
+  'number-float': NumberFieldSchema,
+  'number-integer': NumberFieldSchema,
+  'image-picker': ImagePickerFieldSchema,
+  divider: DividerFieldSchema,
+}
+
+export const MarketingActivityExtensionSchema = BaseSchema.extend({
```

Comment:
The schema is really complex, I assume it's been throughly tested!

Question: have you explored the option to define this in the backend using contracts? a "contract" is a JSON schema defined in the backend that the CLI will fetch to validate the extension, instead of hardcoding the schema in the CLI.

You can see an example of an extension using contracts here: https://github.com/Shopify/shopify/blob/6896fe49dd46fcb7b9c8b99f0fb645f9e7bfd869/components/content/metafields/config/initializers/app_modules.rb#L5